File: IOMAG5_0506/main_sajat.py
import pymongo as mongo

client = mongo.MongoClient("mongodb://localhost:27017")
db = client["konyvtar"]

konyv_coll = db["konyv"]
tag_coll = db["tagok"]

konyv_coll.delete_many({})
tag_coll.delete_many({})

# --- 1. Adatok feltöltése ---

konyvek_adatok = [
    {
        "_id": "k1",
        "cim": "Clean Code",
        "szerzo": "Robert C. Martin",
        "kiado": "Prentice Hall",
        "kiadas_eve": 2012,
        "nyelv": "Angol",
        "mufajok": ["programozás", "szoftverfejlesztés"]
    },
    {
        "_id": "k2",
        "cim": "Songs of Fire and Ice",
        "szerzo": "George R.R. Martin",
        "kiado": "Teszt Kiadó",
        "kiadas_eve": 2010,
        "nyelv": "Angol",
        "mufajok": ["fantasy", "kaland"]
    },
    {
        "_id": "k3",
        "cim": "Dune",
        "szerzo": "Frank Herbert",
        "kiado": "Teszt Kiadó",
        "kiadas_eve": 2013,
        "nyelv": "Angol",
        "mufajok": ["sci-fi", "kaland"]
    }
]

tagok_adatok = [
    {
        "_id": "t1",
        "nev": "Teszt Elek",
        "lakcim": {
            "iranyitoszam": "5600",
            "utca": "Alma",
            "hazszam": "1"
        },
        "kor": 28,
        "kedvenc_kiado": "Prentice Hall",
        "erdeklodesek": ["programozás", "sci-fi"]
    },
    {
        "_id": "t2",
        "nev": "Teszt Tímea",
        "lakcim": {
            "iranyitoszam": "3300",
            "utca": "Körte",
            "hazszam": "1"
        },
        "kor": 35,
        "kedvenc_kiado": "Teszt Kiadó",
        "erdeklodesek": ["fantasy", "történelem"]
    },
    {
        "_id": "t3",
        "nev": "Kiss Péter",
        "lakcim": {
            "iranyitoszam": "1011",
            "utca": "Fő",
            "hazszam": "12"
        },
        "kor": 41,
        "kedvenc_kiado": "Prentice Hall",
        "erdeklodesek": ["fantasy", "történelem", "sci-fi"]
    }
]

konyv_coll.insert_many(konyvek_adatok)
tag_coll.insert_many(tagok_adatok)
print("Adatok feltöltve.")

# --- 2. Összes könyv kiírása ---

for konyv in konyv_coll.find():
    print(konyv)

print("------------------------------------------------------------")

# --- 3. Összes tag kiírása ---

for tag in tag_coll.find():
    print(tag)

print("------------------------------------------------------------")

# --- 4. Egy könyv lekérdezése ID alapján ---

konyv_k1 = konyv_coll.find_one({"_id": "k1"})
print(konyv_k1)

print("------------------------------------------------------------")

# --- 5. 2011 után kiadott könyvek szűrése ---

for konyv in konyv_coll.find({"kiadas_eve": {"$gte": 2011}}):
    print(konyv)

print("-------------------------------------------------------------")

# --- 6. Tagok átlagos életkora (aggregáció) ---

pipeline_avg = [
    {
        "$group": {
            "_id": None,
            "atlagKor": {"$avg": "$kor"}
        }
    }
]
atlag_eredmeny = list(tag_coll.aggregate(pipeline_avg))
atlag = atlag_eredmeny[0]['atlagKor']
print(f"A tagok átlagos életkora: {atlag:.2f} év")

print("--------------------------------------------------------")

# --- 7. Tagok szűrése érdeklődés alapján + könyveik (lookup/JOIN) ---

pipeline = [
    {
        "$match": {
            "erdeklodesek": "programozás"
        }
    },
    {
        "$lookup": {
            "from": "konyv",
            "localField": "kedvenc_kiado",
            "foreignField": "kiado",
            "as": "konyv_adatok"
        }
    }
]

for doc in tag_coll.aggregate(pipeline):
    konyvek_str = ", ".join([k['cim'] for k in doc['konyv_adatok']]) if doc['konyv_adatok'] else "Nincs adat"
    print(f"Tag: {doc['nev']}----------------- Könyvek: {konyvek_str}")

# --- 8. Módosítás: k1-es könyv kiadási évének frissítése ---

print("---------------------------Update.-------------------------------")
print("\n--- k1-es ID könyv módosítása ---")
konyv_coll.update_one(
    {"_id": "k1"},
    {"$set": {"kiadas_eve": 2020}}
)

for konyv in konyv_coll.find():
    print(konyv)

print("--------------------------------------------------------")

# --- 9. Egy tag törlése (t2) ---

tag_coll.delete_one({"_id": "t2"})
print("t2-es tag törölve:")
for t in tag_coll.find():
    print(t)

print("--------------------------------------------------------")

# --- 10. 30 év alatti tagok törlése ---

torles_eredmeny = tag_coll.delete_many(
    {"kor": {"$lt": 30}}
)
print(f"Törölt: {torles_eredmeny.deleted_count}")
for t in tag_coll.find():
    print(t)

File: IOMAG5_0429_2/IOMAG5_mongoAPI_java/src/main/java/org/example/Main_sajat.java
package org.example;

import com.mongodb.client.*;
import org.bson.Document;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static com.mongodb.client.model.Accumulators.avg;
import static com.mongodb.client.model.Aggregates.*;
import static com.mongodb.client.model.Filters.*;

public class Main_sajat {
    public static void main(String[] args) {

        String connectionString = "mongodb://localhost:27017";
        try (MongoClient mongoClient = MongoClients.create(connectionString)) {
            MongoDatabase database = mongoClient.getDatabase("KonyvtarDB");

            MongoCollection<Document> konyvColl = database.getCollection("konyv");
            MongoCollection<Document> tagColl = database.getCollection("tagok");

            // --- 2a) 3 könyv kézi beszúrása (_id > k3) ---

            Document k4 = new Document("_id", "k4")
                    .append("cim", "A Gyűrűk Ura")
                    .append("szerzo", "J.R.R. Tolkien")
                    .append("kiado", "Allen & Unwin")
                    .append("kiadas_eve", 2001)
                    .append("nyelv", "Magyar")
                    .append("mufajok", Arrays.asList("fantasy", "kaland"));

            Document k5 = new Document("_id", "k5")
                    .append("cim", "1984")
                    .append("szerzo", "George Orwell")
                    .append("kiado", "Secker & Warburg")
                    .append("kiadas_eve", 1949)
                    .append("nyelv", "Angol")
                    .append("mufajok", Arrays.asList("disztópia", "sci-fi"));

            Document k6 = new Document("_id", "k6")
                    .append("cim", "Egri csillagok")
                    .append("szerzo", "Gárdonyi Géza")
                    .append("kiado", "Dante Kiadó")
                    .append("kiadas_eve", 1901)
                    .append("nyelv", "Magyar")
                    .append("mufajok", Arrays.asList("történelmi", "kaland"));

            konyvColl.insertMany(Arrays.asList(k4, k5, k6));

            // --- 2b) JSON fájlok beolvasása ---

            String konyvekRaw = Files.readString(Paths.get("konyvek_11.json"));
            List<Document> konyvekList = Document.parse(konyvekRaw).getList("TempList", Document.class);
            List<Document> formattedKonyvek = konyvekList.stream()
                    .map(doc -> {
                        doc.put("_id", doc.getString("kkod"));
                        doc.remove("kkod");
                        if (doc.containsKey("kiadas_eve")) {
                            doc.put("kiadas_eve", Integer.parseInt(doc.getString("kiadas_eve")));
                        }
                        return doc;
                    })
                    .collect(Collectors.toList());
            konyvColl.insertMany(formattedKonyvek);

            String tagokRaw = Files.readString(Paths.get("tagok_11.json"));
            List<Document> tagokList = Document.parse(tagokRaw).getList("TempList", Document.class);
            List<Document> formattedTagok = tagokList.stream()
                    .map(doc -> {
                        doc.put("_id", doc.getString("tkod"));
                        doc.remove("tkod");
                        if (doc.containsKey("kor")) {
                            doc.put("kor", Integer.parseInt(doc.getString("kor")));
                        }
                        if (doc.containsKey("tagdij")) {
                            doc.put("tagdij", Integer.parseInt(doc.getString("tagdij")));
                        }
                        return doc;
                    })
                    .collect(Collectors.toList());
            tagColl.insertMany(formattedTagok);

            // --- 3a) Összes könyv lekérdezése ---

            System.out.println("--- a) Összes könyv lekérdezése ---");
            konyvColl.find().into(new ArrayList<>())
                    .forEach(doc -> System.out.println(doc.toJson()));

            // --- 3b) Könyv lekérdezése (_id: k3) ---

            System.out.println("\n--- b) Könyv lekérdezése (kkod: k3) ---");
            Document konyvK3 = konyvColl.find(eq("_id", "k3")).first();
            if (konyvK3 != null) {
                System.out.println(konyvK3.toJson());
            }

            // --- 3c) Tagok, akik idősebbek 35 évnél ---

            System.out.println("\n--- c) Tagok, akik idősebbek 35 évnél ---");
            tagColl.find(gt("kor", 35)).into(new ArrayList<>())
                    .forEach(doc -> System.out.println(doc.toJson()));

            // --- 3d) 41 éves tag és kedvenc könyvei (JOIN) ---

            System.out.println("\n--- d) Tag (41 éves) és kedvenc kiadójának könyvei (JOIN) ---");
            tagColl.aggregate(Arrays.asList(
                    match(eq("kor", 41)),
                    lookup("konyv", "kedvenc_kiado", "kiado", "konyv_info")
            )).into(new ArrayList<>())
                    .forEach(doc -> System.out.println(doc.toJson()));

            // --- 3e) Tagok átlagos életkora ---

            System.out.println("\n--- e) Tagok átlagos életkora ---");
            Document atlag = tagColl.aggregate(Arrays.asList(
                    group(null, avg("atlagKor", "$kor"))
            )).first();
            System.out.println("Átlagéletkor: " + atlag.get("atlagKor"));

            // --- 4a) k2 könyv címének módosítása ---

            konyvColl.updateOne(
                    eq("_id", "k2"),
                    new Document("$set", new Document("cim", "A Tűz és Jég Dala"))
            );

            // --- 4b) k4 könyv törlése ---

            konyvColl.deleteOne(eq("_id", "k4"));

            // --- 4c) 30 év alatti tagok törlése ---

            tagColl.deleteMany(lt("kor", 30));

        } catch (Exception e) {
            System.err.println("Hiba történt: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
